File: src/ios_loc/gpx.py
"""Parse GPX files into a plain list of coordinates.

Pure parsing, no I/O: like `routing.decode_polyline`, this module never
touches the filesystem or network -- callers read the file and pass the text
in.
"""

from __future__ import annotations

from xml.etree import ElementTree

from ios_loc.path import Coord


def _local_name(tag: str) -> str:
    """Strip a GPX namespace off an element tag, e.g. '{...}trkpt' -> 'trkpt'."""
    return tag.rsplit("}", 1)[-1]


def _points(root: ElementTree.Element, tag: str) -> list[Coord]:
    points: list[Coord] = []
    for elem in root.iter():
        if _local_name(elem.tag) != tag:
            continue
        lat_raw, lon_raw = elem.get("lat"), elem.get("lon")
        if lat_raw is None or lon_raw is None:
            raise ValueError(f"<{tag}> is missing a lat or lon attribute")
        try:
            points.append((float(lat_raw), float(lon_raw)))
        except ValueError as exc:
            raise ValueError(
                f"<{tag}> has a non-numeric lat/lon: {lat_raw!r}, {lon_raw!r}"
            ) from exc
    return points


def parse_gpx(text: str) -> list[Coord]:
    """Extract a route's coordinates from GPX XML.

    Prefers track points (`<trkpt>`, concatenated across every `<trkseg>` in
    document order -- a multi-segment track is one continuous path here).
    Falls back to route points (`<rtept>`) if there is no track, then to
    waypoints (`<wpt>`) if there is neither: a recorded track is the most
    literal description of a path a GPX file can hold, so it wins whenever
    it's present.

    Raises `ValueError` for anything that isn't parseable GPX: invalid XML, a
    point missing `lat`/`lon`, or fewer than 2 points found in total.
    """
    try:
        root = ElementTree.fromstring(text)
    except ElementTree.ParseError as exc:
        raise ValueError(f"not valid XML: {exc}") from exc

    for tag in ("trkpt", "rtept", "wpt"):
        points = _points(root, tag)
        if points:
            if len(points) < 2:
                raise ValueError(
                    f"GPX file has only {len(points)} <{tag}> point(s); a route needs at least 2"
                )
            return points

    raise ValueError("GPX file has no <trkpt>, <rtept>, or <wpt> points")

File: src/ios_loc/web/ui/src/lib/gpx.ts
import type { LatLon } from "@/api/types"
import type { DraftRoute } from "@/lib/draft"

// A regex scanner rather than DOMParser: vitest's frontend suite is
// pure-logic-only (node environment, no jsdom -- see vitest.config.ts), so
// anything reachable from a test can't depend on a browser DOM API.

function attr(tagAttrs: string, name: string): string | null {
  const match = tagAttrs.match(new RegExp(`\\b${name}\\s*=\\s*(?:"([^"]*)"|'([^']*)')`))
  if (!match) return null
  return match[1] ?? match[2] ?? null
}

function pointsFor(text: string, tag: string): LatLon[] {
  const points: LatLon[] = []
  const tagRe = new RegExp(`<${tag}\\b([^>]*?)/?>`, "g")
  for (const match of text.matchAll(tagRe)) {
    const tagAttrs = match[1]
    const latRaw = attr(tagAttrs, "lat")
    const lonRaw = attr(tagAttrs, "lon")
    if (latRaw === null || lonRaw === null) {
      throw new Error(`<${tag}> is missing a lat or lon attribute`)
    }
    const lat = Number.parseFloat(latRaw)
    const lon = Number.parseFloat(lonRaw)
    if (!Number.isFinite(lat) || !Number.isFinite(lon)) {
      throw new Error(
        `<${tag}> has a non-numeric lat/lon: ${JSON.stringify(latRaw)}, ${JSON.stringify(lonRaw)}`,
      )
    }
    if (lat < -90 || lat > 90 || lon < -180 || lon > 180) {
      throw new Error(`<${tag}> out of range: latitude ${lat}, longitude ${lon}`)
    }
    points.push([lat, lon])
  }
  return points
}

/**
 * Parse GPX XML into a `DraftRoute`. Mirrors `ios_loc.gpx.parse_gpx`'s
 * priority: track points (`<trkpt>`, every `<trkseg>` concatenated in
 * document order) win over route points (`<rtept>`), which win over waypoints
 * (`<wpt>`) -- a recorded track is the most literal description of a path a
 * GPX file can hold, so it wins whenever present.
 */
export function gpxToRoute(text: string, literal: boolean): DraftRoute | { error: string } {
  if (!/<gpx[\s>]/i.test(text)) {
    return { error: "not a GPX file (no <gpx> root element found)" }
  }
  try {
    for (const tag of ["trkpt", "rtept", "wpt"]) {
      const points = pointsFor(text, tag)
      if (points.length === 0) continue
      if (points.length < 2) {
        return {
          error: `GPX file has only ${points.length} <${tag}> point(s); a route needs at least 2`,
        }
      }
      return { waypoints: points, name: null, literal }
    }
  } catch (error) {
    return { error: error instanceof Error ? error.message : String(error) }
  }
  return { error: "GPX file has no <trkpt>, <rtept>, or <wpt> points" }
}
